newsrevenuehub/rev-engine
# NRH-242  | facebook pixel integration

DON’T MERGE ME IN TO DEVELOP UNTIL:

#### What's this PR do?

User-wise...

- Hub admins can add a facebook pixel id to a revenue program
- Donation pages for that revenue program will send a pageview event to facebook pixel
- Thank you pages for that revenue program will send a pageview event to facebook pixel
- When a user successfully donates, "Donate" and "Purchase" events will be sent to Facebook Pixel

Code-wise...

- Refactors SPA-side analytics infrastructure to use React context for initializing and sharing a single analytics instance. This was necessary in order to make FB Pixel conversion-tracking unit testable 
- Refactors page tracking to use a `<TrackPageView>` component who gets analytics instance from context, and doesn't need to be smart about getting child pages to get analytics instance in correct state.
- Adds the concept of a "conversion". A conversion is something a contributor does, and it can trigger other events. In our case, it triggers analytics events to be fired on FB pixel. In the future other analytics providers could track this as well.
- Adds new npm commands: `npm run cypress:run:component` and `npm run cypress:open:component` which are for running Cypress-based React component unit tests. This is used for testing the `trackConversion` global function (*global* in the sense of being provided via context), which is what calls fb pixel.


#### How should this be manually tested?

- Add valid facebook pixel id to a revenue program
- Open a browser tab, and open developer tools
- In same browser tab, go to a live donation page for the rev program
- Confirm a pageview event fires to facebook in network tab of dev tools. The url will be like https://connect.facebook.net/signals/config/<ORG_FACEBOOK_PIXEL_ID>?v=2.9.44&r=stable (substituting a real value for ORG_FACEBOOK_PIXEL_ID).
- Successfully complete a donation and get sent to the thank you page
- Confirm a pageview event fires to facebook in network tab of dev tools. The url will be like https://connect.facebook.net/signals
- Confirm "Donate" and "Purchase" events fired by looking for two requests to https://www.facebook.com/tr/ in the network tab of developer tools. For donate, the query param "ev" will be "Donate", and for purchase, "ev" will be "Purchase"
- Additionally, can confirm from inside Facebook Pixel interface, if set up a live instance

#### Do any changes need to be made before deployment to staging? production? (adding environment variables, for example)?

No
